src/extractor/edge_based_graph_factory.cpp

```diff
@@ -638,6 +595,475 @@ void EdgeBasedGraphFactory::GenerateEdgeExpandedEdges(
+    };
+    const auto getRight = [&turn_candidates](std::size_t index)
+    {
+        return (index + turn_candidates.size() - 1) % turn_candidates.size();
+    };
+    const auto &candidate = turn_candidates[turn_index];
+    const EdgeData &in_data = m_node_based_graph->GetEdgeData(via_eid);
+    const EdgeData &out_data = m_node_based_graph->GetEdgeData(candidate.eid);
+    const auto &candidate_to_the_left = turn_candidates[getLeft(turn_index)];
+
+    const auto &candidate_to_the_right = turn_candidates[getRight(turn_index)];
+
+    const auto hasValidRatio = [](const TurnCandidate &left, const TurnCandidate &center,
+                                  const TurnCandidate &right)
+    {
+        auto angle_left = (left.angle > 180) ? angularDeviation(left.angle, STRAIGHT_ANGLE) : 180;
+        auto angle_right =
+            (right.angle < 180) ? angularDeviation(right.angle, STRAIGHT_ANGLE) : 180;
+        auto self_angle = angularDeviation(center.angle, STRAIGHT_ANGLE);
+        return angularDeviation(center.angle, STRAIGHT_ANGLE) < NARROW_TURN_ANGLE &&
+               ((center.angle < STRAIGHT_ANGLE)
+                    ? (angle_right > self_angle && angle_left / self_angle > DISTINCTION_RATIO)
+                    : (angle_left > self_angle && angle_right / self_angle > DISTINCTION_RATIO));
+    };
+    // only valid turn
+
+    return turn_candidates.size() == 1 ||
+           // only non u-turn
+           (turn_candidates.size() == 2 &&
+            candidate_to_the_left.instruction == TurnInstruction::UTurn) || // nearly straight turn
+           angularDeviation(candidate.angle, STRAIGHT_ANGLE) < MAXIMAL_ALLOWED_NO_TURN_DEVIATION ||
+           hasValidRatio(candidate_to_the_left, candidate, candidate_to_the_right) ||
+           (in_data.name_id != 0 && in_data.name_id == out_data.name_id &&
+            angularDeviation(candidate.angle, STRAIGHT_ANGLE) < NARROW_TURN_ANGLE / 2);
+}
+
+std::vector<EdgeBasedGraphFactory::TurnCandidate>
+EdgeBasedGraphFactory::suppressTurns(EdgeID via_eid,
+                                     std::vector<TurnCandidate> turn_candidates) const
+{
+    // remove invalid candidates
+    BOOST_ASSERT_MSG(std::is_sorted(turn_candidates.begin(), turn_candidates.end(),
+                                    [](const TurnCandidate &left, const TurnCandidate &right)
+                                    {
+                                        return left.angle < right.angle;
+                                    }),
+                     "Turn Candidates not sorted by angle.");
+    const auto end = std::stable_partition(turn_candidates.begin(), turn_candidates.end(),
+                                           [](const TurnCandidate &candidate)
+                                           {
+                                               return candidate.valid;
+                                           });
+    turn_candidates.resize(std::distance(turn_candidates.begin(), end));
```

Comment:
I don't understand the need for stable_partition+resize here. What about erase+remove_if?